Tim N. -- Activity 2 & 3

iefomit opened the pull request:
## Overview
This PR will simulate the robot using the new subsystem (activity 2), as well as changed the drive type to Arcade Drive (activity 3), and complete three autonomous commands -- drive forward, drive square, & Bang-Bang drive -- (activity).

## Activity Finished
Completed both Activity 2 and Activity 3

## Tests Ran
Simulated robot using subsystems DriveTrain.java & DriveSub.java (created 2.3). In activity 3, ran the 3 of the 3 autonomous tasks successfully from Shuffleboard.

iefomit committed:
Initialize motors

iefomit committed:
Update Drivetrain.java

iefomit committed:
Merge branch 'main' into origin/Tim

iefomit committed:
Finished Activity 1

iefomit committed:
Finished 100% Activity 1

iefomit committed:
Merge branch 'main' into origin/Tim

iefomit committed:
Finished 101% Activity 1

Set the idle mode on both motors to brake

iefomit committed:
Finished 2.1

Completed Drivetrain Sim

iefomit committed:
Finishing 2.2

iefomit committed:
Finished 2.2

iefomit committed:
Finishing 2.3

Mad new subsystem

iefomit committed:
Fixing Things 2.3

iefomit committed:
Finished 2.3

iefomit committed:
Finished 2.4

iefomit committed:
Finished 101% Activity 2.4

iefomit committed:
Finished 3.1

Arnav814 commented on src/main/java/frc/robot/subsystems/DriveSub.java:
```diff
@@ -0,0 +1,102 @@
+        if (isReal) {
+            motor.set(speed);
+        } else {
+            armSim.setInput(speed * 12.0); // multiply robot voltage
+            armSim.update(0.02); // update the simulation
```
Use the appropriate constants (Constants.ROBOT_VOLTAGE and Constants.LOOP_TIME) instead of typing the numbers directly. The constants communicate more about what the code actually does.

Arnav814 commented on src/main/java/frc/robot/subsystems/DriveSub.java:
```diff
@@ -0,0 +1,102 @@
+    private Mechanism2d mech2d;
+    private MechanismLigament2d wheelLigament;
+    
+    private boolean isReal = false; // check is real or simulated
```
Actually check this with Robot.isReal() instead of passing it in. Passing a value like this around is annoying and hard to maintain.

rostovets committed:
Finished 3.2

rostovets committed:
Autonomous Square

Made an autonomous square program that moves the robot in a counter-clockwise square movement that can be launched from Shuffleboard in simulation

rostovets committed:
Finished 3.3

rostovets committed:
debug activity 2

Kyle-Eldridge commented on src/main/java/frc/robot/commands/AutoTurnCommand.java:
```diff
@@ -0,0 +1,48 @@
+        double angleTurned = currentAngle.minus(startAngle).getDegrees();
+
+        // Stop turning once we've turned the desired angle (negative for left)
+        return Math.abs(angleTurned) >= angle;
```
This won't work if angle is negative. Use Math.abs(angle).

Kyle-Eldridge commented on src/main/java/frc/robot/commands/AutoTurnCommand.java:
```diff
@@ -0,0 +1,48 @@
+public class AutoTurnCommand extends Command {
+    private final Drivetrain drivetrain;
+    private final double angle; // Angle to turn (positive for left turn)
+    private final Timer timer = new Timer();
```
This command doesn't use the Timer.

Kyle-Eldridge commented on src/main/java/frc/robot/commands/BangBangDriveCommand.java:
```diff
@@ -0,0 +1,42 @@
+    @Override
+    public boolean isFinished() {
+        double currentPosition = drivetrain.getAveragePosition();
+        return Math.abs(currentPosition - setpoint) < 0.05; // finish within 5 cm of setpoint
```
Try lowering the tolerance to get to get closer to the setpoint.

Kyle-Eldridge commented on src/main/java/frc/robot/commands/BangBangSubsystemCommand.java:
```diff
@@ -0,0 +1,37 @@
+    public void execute() {
+        double currentPosition = subsystem.getPosition();
+        if (currentPosition < setpoint) {
+            subsystem.motor.set(power); // drive forward
```
Use the method you made to set the motor. Never set its power directly. It should also be private.

Kyle-Eldridge commented on src/main/java/frc/robot/commands/BangBangSubsystemCommand.java:
```diff
@@ -0,0 +1,37 @@
+    @Override
+    public boolean isFinished() {
+        double currentPosition = subsystem.getPosition();
+        return Math.abs(currentPosition - setpoint) < 0.05; // adjust tolerance as needed
```
This will probably never finish because this is in degrees. It won't be able to get within one twentieth of a degree with Bang-Bang control.

Kyle-Eldridge replied:
Maybe change it to 0.5 or 1 degree instead.

Kyle-Eldridge commented on src/main/java/frc/robot/commands/BangBangRotateCommand.java:
```diff
@@ -0,0 +1,9 @@
+
+public class BangBangRotateCommand extends BangBangSubsystemCommand {
+    public BangBangRotateCommand(DriveSub subsystem, double rotations) {
+        super(subsystem, rotations * 360); // convert rotations to degrees
```
This is supposed to rotate the subsystem by a certain number of rotations from its current position, so that it will turn froward the same amount every time you run it.

iefomit replied:
Did my commit fix it?

Kyle-Eldridge replied:
Do not resolve conversations on GitHub. This still won't work. You don't need the setPosition() method in your DriveSub, since it will just reset the encoder to the input value. I added some more detailed instructions to step 3.3.12 on the Activities document saying how to do this.

Kyle-Eldridge replied:
Why did you delete everything, including the class definition?

iefomit replied:
hopefully fixed it, check my commit

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/DriveSub.java:
```diff
@@ -0,0 +1,103 @@
+
+public class DriveSub extends SubsystemBase {
+
+    public CANSparkMax motor;
```
It's a TalonFX, not a CANSparkMax.

Kyle-Eldridge replied:
Do not resolve conversations before you make the requested changes. Your code still has a CANSparkMax as the motor, not a TalonFX. It's usually better to not resolve any conversations and to instead wait for the person who made the comment to resolve it.

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/DriveSub.java:
```diff
@@ -0,0 +1,103 @@
+            motor.set(speed);
+        } else {
+            armSim.setInput(Constants.ROBOT_VOLTAGE); // multiply robot voltage
+            armSim.update(Constants.LOOP_TIME); // update the simulation
```
Like the comment says, you need to multiply ROBOT_VOLTAGE by the speed, instead of always setting to ROBOT_VOLTAGE.

Kyle-Eldridge replied:
This is still not fixed.

iefomit replied:
Just fixed it)

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/DriveSub.java:
```diff
@@ -0,0 +1,103 @@
+        if (RobotBase.isReal()) {
+            return motor.getEncoder().getPosition();
+        } else {
+
```
This is in rotations, while the code for sim will return degrees. Change one of them so the units match.

iefomit replied:
I'm not sure if my recent edit in "BangBangRotateCommand.java fixed it.

Kyle-Eldridge replied:
You need to fix it here. This method should return the same units if the robot is real or simulated, which it currently does not. You can choose if you want everything in rotations, radians, or degrees, but it should be the same. Otherwise, your commands would work in sim but not in real life.

iefomit replied:
hopefully fixed it, check my commit

iefomit replied:
*I was planning everything to be in rotations*

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/DriveSub.java:
```diff
@@ -0,0 +1,103 @@
+        if (RobotBase.isReal()) {
+            setSpeed(0.05); // temp. speed setting for real robot
+        } else {
+            setSpeed(.005); // slower speed for sim
```
Now that you are controlling this using commands, you don't need this code to set the speed anymore.

iefomit replied:
Should I delete the whole periodic method?

Kyle-Eldridge replied:
No, just the code to set the motors. Leave the wheelLigament.setAngle(getPosition()); part.

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -16,26 +25,39 @@ public class Drivetrain extends SubsystemBase {
+  DifferentialDrivetrainSim driveSim;
+
   // TODO 2.2.1: Create gyro (AHRS)
+  AHRS ahrs = new AHRS(SPI.Port.kMXP);
```
private

Kyle-Eldridge replied:
I meant all instance variables in Drivetrain, not just ahrs. I didn't want to add separate comments for every single variable.

iefomit replied:
fixed it

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -16,26 +25,39 @@ public class Drivetrain extends SubsystemBase {
+
+    if (!RobotBase.isReal()) {
+    driveSim = new DifferentialDrivetrainSim(DriveConstants.DRIVETRAIN_PLANT, DriveConstants.MOTOR, DriveConstants.GEAR_RATIO, DriveConstants.TRACK_WIDTH, DriveConstants.WHEEL_DIAMETER / 2, DriveConstants.MEASUREMENT_STD_DEVS);
+    drivePose = new DifferentialDrivePoseEstimator(driveKinematics, getGyroAngle(), getLeftPosition(), getRightPosition(), new Pose2d());
```
drivePose should exist for both real and simulated robots. Define it one line lower, after the }, to fix this.

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -46,11 +68,19 @@ public Drivetrain() {
+    if (RobotBase.isReal()) {
+      
+    } else {
+      driveSim.update(Constants.LOOP_TIME);
```
You don't need the empty if statement. Either use ! or RobotBase.isSimulation() instead.

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -62,10 +92,16 @@ public void periodic(){
-    // TODO 1.2.1: Implement tankDrive
+    leftMotor1.set(leftPower * .25);
+    rightMotor1.set(rightPower * .25);
 
```
You don't need to set the motor speed here since you set them a few lines later.

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -76,47 +112,65 @@ public void tankDrive(double leftPower, double rightPower) {
+ public double getLeftPosition(){
+
+    if (RobotBase.isReal()){
+      return leftMotor1.getEncoder().getPosition();
```
This is in motor rotations. You need to divide by the gear ratio and multiply by the wheel circumference to convert to meters.

Kyle-Eldridge replied:
Same thing for getRightPosition()

iefomit replied:
fixed it

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/Drivetrain.java:
```diff
@@ -76,47 +112,65 @@ public void tankDrive(double leftPower, double rightPower) {
-  public double getLeftSpeed(){
-    return 0;
-  }
-  public double getRightSpeed(){
```
You will need this methods later. Don't delete them.

Kyle-Eldridge commented on src/main/java/frc/robot/util/ShuffleBoard/Tabs/SubsystemTab.java:
```diff
@@ -1,23 +1,30 @@
 public class SubsystemTab extends ShuffleBoardTabs {
     // TODO 2.3.11: Create variable for subsystem
+    
+    public DriveSub driveSub;
```
private

Kyle-Eldridge commented on src/main/java/frc/robot/util/ShuffleBoard/Tabs/SubsystemTab.java:
```diff
@@ -1,23 +1,30 @@
+
 
         // TODO 3.3.13: Add command buttons
+        tab.add("Spin Motor", new BangBangRotateCommand(driveSub, 5)); // spin motor 5 rotations
```
Add your BangBangSubsystemCommand, either here or in AutoTab.

Kyle-Eldridge replied:
Why does it set it to 1 degree?

Kyle-Eldridge replied:
Maybe change it to 90 or 180 degrees, or another angle that is far enough to notice the change.

iefomit replied:
I was trying to change everything to *rotations* (360 degrees), I changed the *Bang Bang Command*, which hopefully runs in rotations.

Kyle-Eldridge reviewed (changes requested):
Always test all of your code before making a PR. Some of this code doesn't work yet.

iefomit committed:
Fixing Activity 3

johaj10 commented on src/main/java/frc/robot/Robot.java:
```diff
@@ -20,38 +23,47 @@
     // TODO 2.3.10: Create your subsystem
     
-    shuffleboard = new ShuffleBoardManager(drive);
+    shuffleboard = new ShuffleBoardManager(drive, driveSub);
```
Your constructor definition in the ShuffleboardManager only takes in one parameter.

Kyle-Eldridge replied:
It takes 2:
```java
public ShuffleBoardManager(Drivetrain drive, DriveSub driveSub){
```

johaj10 reviewed (changes requested):
Address the changes that Kyle and I requested, then we can review again. You're doing a good job!

Kyle-Eldridge commented:
You also need to update from main before we approve this. It might break your cod e or cause merge conflicts that you need to fix, but you will have to do this often in build season, so it's good practice.

iefomit committed:
Fixing more Activity 3

iefomit committed:
fixing

iefomit committed:
Still fixing (WILL change Rotate Command)

iefomit committed:
fixing more

iefomit committed:
maybe done

Kyle-Eldridge commented on src/main/java/frc/robot/commands/BangBangRotateCommand.java:
```diff
@@ -0,0 +1,27 @@
+
+        public double getPoint() {
+        return rotations;
+        }
```
Unused method

iefomit replied:
I deleted it. Not sure if it's needed

Kyle-Eldridge commented on src/main/java/frc/robot/commands/BangBangRotateCommand.java:
```diff
@@ -0,0 +1,27 @@
+
+    @Override
+    public void initialize() {
+      setPoint(subsystem.getPosition() + rotations * 360);
```
Since you changed everything in DriveSub to rotations, you don't need to multiply by 360 anymore.

iefomit replied:
fixed it

Kyle-Eldridge replied:
It now uses integer division, so 1/360 = 0. To fix this, you can use 1.0/360 or Units.rotationsToDegrees().

Kyle-Eldridge commented on src/main/java/frc/robot/commands/BangBangSubsystemCommand.java:
```diff
@@ -0,0 +1,41 @@
+public class BangBangSubsystemCommand extends Command{
+    private final DriveSub subsystem;
+    private final double setpoint;
+    private final double power = 0.5;
```
Power of 0.5 is a lot for the subsystem. You need to test different values, but 0.01 might be good.

iefomit replied:
fixed it

Kyle-Eldridge commented on src/main/java/frc/robot/commands/BangBangSubsystemCommand.java:
```diff
@@ -0,0 +1,41 @@
+
+    public void setPoint(double setpoint) {
+
+    }
```
Make this method work.

iefomit replied:
should work now

Kyle-Eldridge commented on src/main/java/frc/robot/commands/BangBangSubsystemCommand.java:
```diff
@@ -0,0 +1,41 @@
+    @Override
+    public boolean isFinished() {
+        double currentPosition = subsystem.getPosition();
+        return Math.abs(currentPosition - setpoint) < .5; // tolerance
```
Now that everything is in rotations, a tolerance of 0.5 rotations is too high. You can divide this by 360 or set it to some arbitrarily lower value.

iefomit replied:
changed the tolerance to 1 / 360

Kyle-Eldridge replied:
It now uses integer division, so 1/360 = 0. To fix this, you can use 1.0/360 or Units.rotationsToDegrees().

Kyle-Eldridge replied:
I think I said the wrong thing. I meant degreesToRotations, not rotationsToDegrees. You can use 1.0/360 or Units.degreesToRotations(1).

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/DriveSub.java:
```diff
@@ -0,0 +1,103 @@
+            motor.set(speed);
+        } else {
+            armSim.setInput(Constants.ROBOT_VOLTAGE * speed); // multiply robot voltage by speed
+            armSim.update(Constants.LOOP_TIME); // update simulation
```
You should update armSim in periodic(), because physics shouldn't stop when you don't set the power.

iefomit replied:
fixed it

Kyle-Eldridge reviewed (changes requested):
This code is almost finished, but you still have some very small changes to make. Also update from main and remove unused imports.

iefomit committed:
finishing activity 3

Kyle-Eldridge commented on src/main/java/frc/robot/subsystems/DriveSub.java:
```diff
@@ -0,0 +1,104 @@
+            motor.set(speed);
+        } else {
+            armSim.setInput(Constants.ROBOT_VOLTAGE * speed); // multiply robot voltage by speed
+            armSim.update(Constants.LOOP_TIME); // update simulation
```
armSim should not be updated here, only in periodic().

Also, update from main. You will get some merge conflicts that you will have to fix, but this is good practice for build season when you will have to do the same thing. The easiest way is to press control+shift+U on GitHub Desktop, or you can find it in one of the menus at the top.

iefomit committed:
Finished Activity 3

iefomit committed:
100% finished activity 3

Fixed issue with subsystem motor not going to setpoint correctly in sim

iefomit committed:
gitignore

iefomit committed:
Updated tolerance check

Kyle-Eldridge commented:
After we finish presenting it today, you can move on to Activity 4. It is a shorter activity, but you will have less class time to work on it because of Rodrigo's presentations.
